File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/property/Position.java
/* Copyright Airship and Contributors */

package com.urbanairship.android.layout.property;

import android.view.Gravity;

import com.urbanairship.json.JsonException;
import com.urbanairship.json.JsonMap;

import androidx.annotation.NonNull;

public class Position {

    @NonNull
    public static final Position CENTER =
            new Position(HorizontalPosition.CENTER, VerticalPosition.CENTER);
    @NonNull
    private final HorizontalPosition horizontal;
    @NonNull
    private final VerticalPosition vertical;

    public Position(@NonNull HorizontalPosition horizontal, @NonNull VerticalPosition vertical) {
        this.horizontal = horizontal;
        this.vertical = vertical;
    }
    @NonNull
    public static Position fromJson(@NonNull JsonMap json) throws JsonException {
        String horizontalJson = json.opt("horizontal").optString();
        String verticalJson = json.opt("vertical").optString();

        HorizontalPosition horizontal = HorizontalPosition.from(horizontalJson);
        VerticalPosition vertical = VerticalPosition.from(verticalJson);

        return new Position(horizontal, vertical);
    }

    @NonNull
    public HorizontalPosition getHorizontal() {
        return horizontal;
    }

    @NonNull
    public VerticalPosition getVertical() {
        return vertical;
    }

    public int getGravity() {
        return Gravity.CENTER | horizontal.getGravity() | vertical.getGravity();
    }
}

File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/property/Position.kt
/* Copyright Airship and Contributors */
package com.urbanairship.android.layout.property

import android.view.Gravity
import com.urbanairship.json.JsonException
import com.urbanairship.json.JsonValue

public class Position public constructor(
    @JvmField public val horizontal: HorizontalPosition,
    @JvmField public val vertical: VerticalPosition
) {

    public fun getGravity(): Int {
        return Gravity.CENTER or horizontal.gravity or vertical.gravity
    }

    public companion object {
        private const val KEY_HORIZONTAL = "horizontal"
        private const val KEY_VERTICAL = "vertical"

        public val CENTER: Position = Position(HorizontalPosition.CENTER, VerticalPosition.CENTER)

        @Throws(JsonException::class)
        public fun fromJson(json: JsonValue): Position {
            val content = json.requireMap()

            return Position(
                horizontal = HorizontalPosition.from(content.require(KEY_HORIZONTAL)),
                vertical = VerticalPosition.from(content.require(KEY_VERTICAL))
            )
        }
    }
}
